From: Mel Gorman <mgorman@suse.de>
Date: Wed, 7 Mar 2012 09:46:35 +0000
Subject: [PATCH] mm: allow PF_MEMALLOC from softirq context
References: Swap over NFS (fate#304949,bnc#747944)
Patch-mainline: no

With the corrent logic, it's possible for a hardware interrupt with BH
disabled to use PFMEMALLOC reserves which is not intended. This patch
uses in_serving_softirq() to only allow softirqs with PFMEMALLOC set to
access the reserves.

Signed-off-by: Mel Gorman <mgorman@suse.de>
---
 mm/page_alloc.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 29d7a44..1683b41 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -2116,10 +2116,11 @@ gfp_to_alloc_flags(gfp_t gfp_mask)
 	if (likely(!(gfp_mask & __GFP_NOMEMALLOC))) {
 		if (gfp_mask & __GFP_MEMALLOC)
 			alloc_flags |= ALLOC_NO_WATERMARKS;
-		else if (!in_irq() && (current->flags & PF_MEMALLOC))
+		else if (in_serving_softirq() && (current->flags & PF_MEMALLOC))
 			alloc_flags |= ALLOC_NO_WATERMARKS;
 		else if (!in_interrupt() &&
-				unlikely(test_thread_flag(TIF_MEMDIE)))
+				((current->flags & PF_MEMALLOC) ||
+				 unlikely(test_thread_flag(TIF_MEMDIE))))
 			alloc_flags |= ALLOC_NO_WATERMARKS;
 	}
 
